internal/transport/grpchttp2/http2bridge_test.go

```diff
@@ -0,0 +1,463 @@
+ *
+ */
+
+package grpchttp2
+
+import (
+	"testing"
+
+	"golang.org/x/net/http2/hpack"
+)
+
+type fakeConn struct {
+	WBuf []byte
+	RBuf []byte
+}
+
+func (c *fakeConn) Write(p []byte) (int, error) {
+	c.WBuf = append(c.WBuf, p...)
+	return len(p), nil
+}
+
+func (c *fakeConn) Read(p []byte) (int, error) {
+	n := copy(p, c.RBuf)
+	c.RBuf = c.RBuf[n:]
+	return n, nil
+}
+
+func appendUint32(b []byte, x uint32) []byte {
+	return append(b, byte(x>>24), byte(x>>16), byte(x>>8), byte(x))
+}
+
+func readUint32(b []byte) uint32 {
+	return uint32(b[0])<<24 | uint32(b[1])<<16 | uint32(b[2])<<8 | uint32(b[3])
+}
+
+func readUint24(b []byte) int {
+	return int(b[0])<<16 | int(b[1])<<8 | int(b[2])
+}
+
+func readUint16(b []byte) uint16 {
+	return uint16(b[0])<<8 | uint16(b[1])
+}
+
+func checkWrittenHeader(t *testing.T, name string, hdr []byte, size int, ft FrameType, f Flag, streamID uint32) {
+	if gotSize := readUint24(hdr[0:3]); gotSize != size {
+		t.Errorf("%s(): Size: got %d, want %d", name, gotSize, size)
+	}
+	if gotType := FrameType(hdr[3]); gotType != ft {
+		t.Errorf("%s(): Type: got %#x, want %#x", name, gotType, ft)
+	}
+	if gotFlag := Flag(hdr[4]); gotFlag != f {
+		t.Errorf("%s(): Flags: got %#x, want %#x", name, gotFlag, f)
+	}
+	if gotSID := readUint32(hdr[5:9]); gotSID != streamID {
+		t.Errorf("%s(): StreamID: got %d, want %d", name, gotSID, streamID)
+	}
+}
+
+func checkReadHeader(t *testing.T, hdr *FrameHeader, size int, ft FrameType, f Flag, streamID uint32) {
+	if hdr.Size != uint32(size) {
+		t.Errorf("ReadFrame(): Size: got %d, want %d", hdr.Size, size)
+	}
+	if hdr.Type != ft {
+		t.Errorf("ReadFrame(): Type: got %#x, want %#x", hdr.Type, ft)
+	}
+	if hdr.Flags != f {
+		t.Errorf("ReadFrame(): Flags: got %#x, want %#x", hdr.Flags, f)
+	}
+	if hdr.StreamID != streamID {
+		t.Errorf("ReadFrame(): StreamID: got %d, want %d", hdr.StreamID, streamID)
+	}
+
+}
+
+func (s) TestHTTP2Bridge_ReadFrame_Data(t *testing.T) {
+	c := &fakeConn{}
+	recvData := "test data"
+	c.RBuf = append(c.RBuf, 0, 0, byte(len(recvData)), byte(FrameTypeData), byte(FlagDataEndStream))
+	c.RBuf = appendUint32(c.RBuf, 1)
+	c.RBuf = append(c.RBuf, []byte(recvData)...)
+
+	f := NewHTTP2FramerBridge(c, c, 0)
+	fr, err := f.ReadFrame()
+	if err != nil {
+		t.Fatalf("ReadFrame(): %v", err)
+	}
+
+	h := fr.Header()
+	checkReadHeader(t, h, len(recvData), FrameTypeData, FlagDataEndStream, 1)
+	df := fr.(*DataFrame)
+	if string(df.Data) != recvData {
+		t.Errorf("ReadFrame(): Data: got %q, want %q", string(df.Data), recvData)
+	}
+}
+
+func (s) TestHTTP2Bridge_ReadFrame_Headers(t *testing.T) {
+	c := &fakeConn{}
+	recvData := "test data"
+	c.RBuf = append(c.RBuf, 0, 0, byte(len(recvData)), byte(FrameTypeHeaders), byte(FlagHeadersEndHeaders))
+	c.RBuf = appendUint32(c.RBuf, 1)
+	c.RBuf = append(c.RBuf, []byte(recvData)...)
+
+	f := NewHTTP2FramerBridge(c, c, 0)
+	// Done to be able to test reading a single Headers Frame
+	f.framer.ReadMetaHeaders = nil
```

Comment:
In non-test code, if this value is usually not set to `nil`, what is the point of setting it to `nil` here? Could we instead check the meta frame that it returns instead?

Reply:
That makes sense. I will only leave the test for the metaHeaders then.